File: backend/src/main/java/be/lutske/leolegacy/interfaceadapter/rest/RecipeImportDtos.java
package be.lutske.leolegacy.interfaceadapter.rest;

import be.lutske.leolegacy.application.service.IngredientConversion;
import java.util.List;

/**
 * DTOs for the recipe import flow.
 * Field names must match the frontend TypeScript interfaces in api/client.ts.
 */
public final class RecipeImportDtos {

    private RecipeImportDtos() {
    }

    /**
     * Proposed recipe extracted by the AI model.
     * Uses {@code preparation} (joined string) to match the frontend contract.
     */
    public record ProposedRecipeDto(
            String title,
            String description,
            String servings,
            List<String> ingredients,
            String preparation,
            String source,
            List<String> tags,
            Long categoryId,
            String notes,
            List<IngredientConversion> convertedIngredients
    ) {}

    /**
     * Unified response from {@code POST /api/recipes/import}.
     * Always HTTP 200. Status is "COMPLETE" or "NEEDS_MORE_INFO".
     */
    public record ImportExtractionResponse(
            String status,
            String rawModelResponse,
            ProposedRecipeDto proposedRecipe,
            List<String> missingFields,
            List<String> warnings
    ) {}

    /**
     * Request body for confirming/finalizing an imported recipe.
     * Matches the frontend {@code ImportConfirmRequest} interface.
     */
    public record ImportConfirmRequest(
            String rawModelResponse,
            ProposedRecipeDto proposedRecipe,
            UserOverrides userOverrides
    ) {}

    /**
     * User-provided overrides and corrections for the imported recipe.
     */
    public record UserOverrides(
            String title,
            List<String> ingredients,
            String preparation,
            Long categoryId,
            String notes,
            String servings,
            String description
    ) {}
}

File: backend/src/main/kotlin/be/lutske/leolegacy/interfaceadapter/rest/RecipeImportDtos.kt
package be.lutske.leolegacy.interfaceadapter.rest

/**
 * Partial recipe DTO used during the import flow.
 * Fields are nullable because the LLM may not extract everything.
 */
data class ProposedRecipeDto(
    val title: String? = null,
    val description: String? = null,
    val servings: String? = null,
    val ingredients: List<String>? = null,
    val preparation: String? = null,
    val categoryId: Long? = null,
    val notes: String? = null,
    val source: String? = null,
    val tags: List<String>? = null
)

/**
 * Response returned from POST /api/recipes/import after LLM extraction.
 * Always returned as HTTP 200 — the recipe is never auto-saved.
 *
 * Status values:
 * - "COMPLETE": all required fields were extracted successfully
 * - "NEEDS_MORE_INFO": some required fields are missing or uncertain
 * - "ERROR": extraction failed (e.g. image is not a recipe)
 */
data class ImportExtractionResponse(
    val status: String,
    val rawModelResponse: String? = null,
    val proposedRecipe: ProposedRecipeDto,
    val missingFields: List<String> = emptyList(),
    val warnings: List<String> = emptyList()
)

/**
 * Request body for confirming/finalizing an imported recipe.
 */
data class ImportConfirmRequest(
    val rawModelResponse: String? = null,
    val proposedRecipe: ProposedRecipeDto,
    val userOverrides: UserOverrides? = null
)

/**
 * User-provided overrides and corrections for the imported recipe.
 */
data class UserOverrides(
    val title: String? = null,
    val ingredients: List<String>? = null,
    val preparation: String? = null,
    val categoryId: Long? = null,
    val notes: String? = null,
    val servings: String? = null,
    val description: String? = null
)
